Significant rework to hydrogen/tritium combustion.

Pickle-Coding opened the pull request:
<!-- Write **BELOW** The Headers and **ABOVE** The comments else it may not be viewable. -->
<!-- You can view Contributing.MD for a detailed description of the pull request process. -->

## About The Pull Request
Changes hydrogen and tritium combustion gas reactions. Hydrogen/tritium will get consumed at 1 times the burn rate, while oxygen gets consumed at 0.5 times the burn rate, to produce 1 mol of water vapour per burn rate. Significantly changes how the reaction efficiency works. Instead of being an if check on whether there is more oxygen than hydrogen/tritium, then switching between high energy or low energy burn modes, the burn efficiency will instead just be the minimum of half the hydrogen/tritium and 1/20th of the oxygen. The reaction will burn incredibly fast and can end quickly if there is enough oxygen. The energy release of the reactions have been reduced by 90%, which makes them slightly less effective than plasma in total energy gain, but due to their reaction speed, can still be more dangerous and effective for bombs.

As a consequence for the reaction speed, you will collect less tritium when making it as more of the tritium gets consumed before getting scrubbed. The optimal bomb mix has also been changed due to the reaction changes. Generally around 12.85% tritium/hydrogen in a tritium/hydrogen + oxygen mix will give you the best results, filling the tritium or hydrogen to 325.5kPa and oxygen max pressure with portable pump. This does not account of the hot plasma you put in which might change it slightly.
<!-- Describe The Pull Request. Please be sure every change is documented or this can delay review and even discourage maintainers from merging your PR! -->

## Why It's Good For The Game
Fixing the stoichiometry of hydrogen and tritium combustion means there won't be weird gas duplication and deletion when you combust then electrolyse them, and will make power balance easier. The weird binary high energy low energy burn modes are a relic of the ancient past, are confusing to understand and is just generally overcomplicated, and doesn't even achieve its job of preventing singlecaps (it actually allows classic singlecaps to be timed). The new combustion formula is much more intuitive to understand, and combined with the lower energy release, changes its purpose in combustion to be a fast reaction that releases energy quickly, rather than just being a better plasma with a weird if check.
<!-- Argue for the merits of your changes and how they benefit the game, especially if they are controversial and/or far reaching. If you can't actually explain WHY what you are doing will improve the game, then it probably isn't good for the game in the first place. -->

## Changelog

<!-- If your PR modifies aspects of the game that can be concretely observed by players or admins you should add a changelog. If your change does NOT meet this description, remove this section. Be sure to properly mark your PRs to prevent unnecessary GBP loss. You can read up on GBP and it's effects on PRs in the tgstation guides for contributors. Please note that maintainers freely reserve the right to remove and add tags should they deem it appropriate. You can attempt to finagle the system all you want, but it's best to shoot for clear communication right off the bat. -->

:cl:
balance: Changes hydrogen and tritium gas reaction stoichiometry. 1 mol of hydrogen/tritium and 0.5 mols of oxygen getting consumed will produce 1 mol of water vapour.
balance: Hydrogen and tritium burn rate has been changed to be the minimum of half the hydrogen/tritium, or 1/20th of the oxygen, rather than a binary if check on whether there is more oxygen or not.
balance: Hydrogen and tritium energy release has been reduced by 90%.
/:cl:

<!-- Both :cl:'s are required for the changelog to work! You can put your name to the right of the first :cl: if you want to overwrite your GitHub username as author ingame. -->
<!-- You can use multiple of the same prefix (they're only used for the icon ingame) and delete the unneeded ones. Despite some of the tags, changelogs should generally represent how a player might be affected by the changes rather than a summary of the PR's contents. -->

Pickle-Coding committed:
Gaming.

Fixes hydrogen and tritium combustion stochiometry. Hydrogen/tritium combustion will consume 1 mol H2, 0.5 mol O2 to make 1 mol H2O. Removes weird high energy low energy burn if check thing, and instead make the burn rate be based on the minimum of half the hydrogen and 1/20th of the oxygen.

Removes the weird hydrogen combustion scale defines and variables, and increases the hydrogen energy release by a factor of 10 (this is effectively reducing the energy release per mol of hydrogen consumed by 90%, because the hydrogen only got consumed at 10% of the burn rate before this commit, and there was a weird scale thing that multiplied the energy release by 10).

Adjusts the reaction_factors of those gas reactions according to the new changes.

LemonInTheDark commented on code/modules/atmospherics/gasmixtures/reactions.dm:
```diff
@@ -259,27 +259,18 @@
-		burned_fuel = cached_gases[/datum/gas/oxygen][MOLES] / HYDROGEN_BURN_OXY_FACTOR // const must be at least one
-		fire_scale = 1
+	var/burned_fuel = min(cached_gases[/datum/gas/hydrogen][MOLES] / FIRE_HYDROGEN_BURN_RATE_DELTA, cached_gases[/datum/gas/oxygen][MOLES] / (FIRE_HYDROGEN_BURN_RATE_DELTA * HYDROGEN_OXYGEN_FULLBURN), cached_gases[/datum/gas/hydrogen][MOLES], cached_gases[/datum/gas/oxygen][MOLES] * INVERSE(0.5))
+	if(burned_fuel <= 0 || (cached_gases[/datum/gas/hydrogen][MOLES] - burned_fuel < 0) || (cached_gases[/datum/gas/oxygen][MOLES] - burned_fuel * 0.5 < 0)) //Shouldn't produce gas from nothing.
```
don't need the wrapping ()s in the cached gases checks here

LemonInTheDark commented on code/modules/atmospherics/gasmixtures/reactions.dm:
```diff
@@ -259,27 +259,18 @@
-	if(cached_gases[/datum/gas/oxygen][MOLES] < cached_gases[/datum/gas/hydrogen][MOLES] || MINIMUM_HYDROGEN_OXYBURN_ENERGY > (temperature * old_heat_capacity))
-		burned_fuel = cached_gases[/datum/gas/oxygen][MOLES] / HYDROGEN_BURN_OXY_FACTOR // const must be at least one
-		fire_scale = 1
+	var/burned_fuel = min(cached_gases[/datum/gas/hydrogen][MOLES] / FIRE_HYDROGEN_BURN_RATE_DELTA, cached_gases[/datum/gas/oxygen][MOLES] / (FIRE_HYDROGEN_BURN_RATE_DELTA * HYDROGEN_OXYGEN_FULLBURN), cached_gases[/datum/gas/hydrogen][MOLES], cached_gases[/datum/gas/oxygen][MOLES] * INVERSE(0.5))
```
The burn rate here is insane, at least if I'm reading it right. 1/2 hydrogen in one tick seems insane, especially in bomb making. the o2 is a bit better and I get that it's acting as the weight on hydrogen's ankle here but it's still a bit worrying.
Why exactly is it so high?

Pickle-Coding replied:
I want it to be unique by having its extremely fast reaction rate to release energy quickly and explosively, rather than just being plasma but releasing more energy. Its energy release has been decreased, making a fixed amount of a hydrogen and oxygen mix release a similar (burns to around 70,000K water vapour) amount of energy as a plasma oxygen (over 90,000K 25% water vapour 75% co2) mix. I haven't done extensive testing, but have managed to get the anomaly refinery to report an explosion size of 103 (I assume devastation range) with tier 2 cooled t2/o2 and 4,999K plasma TTV.

LemonInTheDark commented on code/modules/atmospherics/gasmixtures/reactions.dm:
```diff
@@ -322,26 +313,16 @@
-		cached_gases[/datum/gas/oxygen][MOLES] -= burned_fuel
-		ASSERT_GAS(/datum/gas/water_vapor, air)
-		cached_gases[/datum/gas/water_vapor][MOLES] += burned_fuel / TRITIUM_BURN_TRIT_FACTOR
+	var/burned_fuel = min(cached_gases[/datum/gas/tritium][MOLES] / FIRE_TRITIUM_BURN_RATE_DELTA, cached_gases[/datum/gas/oxygen][MOLES] / (FIRE_TRITIUM_BURN_RATE_DELTA * TRITIUM_OXYGEN_FULLBURN), cached_gases[/datum/gas/tritium][MOLES], cached_gases[/datum/gas/oxygen][MOLES] * INVERSE(0.5))
+	if(burned_fuel <= 0 || (cached_gases[/datum/gas/tritium][MOLES] - burned_fuel < 0) || (cached_gases[/datum/gas/oxygen][MOLES] - burned_fuel * 0.5 < 0)) //Shouldn't produce gas from nothing.
+		return NO_REACTION
 
+	cached_gases[/datum/gas/tritium][MOLES] -= burned_fuel
+	cached_gases[/datum/gas/oxygen][MOLES] -= burned_fuel * 0.5
+	ASSERT_GAS(/datum/gas/water_vapor, air)
+	cached_gases[/datum/gas/water_vapor][MOLES] += burned_fuel
```
Is this the same reaction as hydrogen, exactly?

Pickle-Coding replied:
Aside from the initial heat capacity difference and radiation pulses, it's basically the same.

Pickle-Coding committed:
Hello.

Removes the excessive parenthesis wrapping from the negative moles sanity checks in the h2fire and tritfire reactions, because they were unnecessary.

LemonInTheDark commented:
Note to self: review this

Pickle-Coding commented:
I have graphed the tritium combustion burn rates for the old and new changes for 2 ticks (for the old one, lets assume the high energy check is true). The total moles of the tritium and oxygen mix for this graph is fixed at 1.

Legend:
- Green: Tritium moles.
- Blue: Oxygen moles.
- Red: Burn rate in the first tick.
- Orange: Burn rate in the second tick.
- Purple: The sum of the burn rates for the first two ticks.

Old:![image](https://user-images.githubusercontent.com/58013024/183412965-5e877680-0f29-4986-a2d5-5fc3e4f580cc.jpeg)

Tritium gets burned at a 10th of this burn rate in the high energy reaction, and oxygen gets burned at the burn rate. The effect scale is 10 times higher in the first half of the graph, making the energy release from the burned fuel 10 times higher.

New:![image](https://user-images.githubusercontent.com/58013024/183416578-ba92228a-95d4-4aa7-87c7-5ee76cc5f602.jpeg)

Tritium gets burned 1:1 here, and oxygen at half the rate. The amount of energy released from tritium combustion can now accurately described by how much tritium or oxygen got combusted, without having to do weird piecewise functions.

Now lets scale the burn rate of the new function by 10, which is how much the energy release define increased, so we can compare the relative energy releases together. Purple will be the combined relative energy release of the new function for the first two ticks, while the old function is red.

![image](https://user-images.githubusercontent.com/58013024/183417717-db8bd36b-0edb-4ed7-ae3c-54eea22bba13.jpeg)

I can not be bothered to scale the old energy release with their weird scale piecewise function, so just imagine for the red line that anything after the first break (or vertical line) is just a horizontal line near 0. This wouldn't matter for optimal TTVs, but it would for failed TTVs.

For optimal TTVs, the peak combined energy release for the new combustion is about 31% higher than the old combustion peak combined energy release. The energy release change wouldn't be a 1:1 change to TTV efficiency, as the stoichiometry changes and therefore the changes to the change of heat capacity can effect the efficiency too, but it should give you an idea of how effective TTVs or singletanks (which would be suicide tanks now that you can't really control a slow reaction anymore) will be. One thing for sure is it will be a lot more friendly and less binary and hostile to people that got their mix slightly wrong.

As for flooding, if supplied with enough oxygen, could get to over 10,000K and potentially melt floors, but will die off quickly whether the floor melts or not, though it is slower if there is an oxygen deficit.

For an adiabatic process such as a canister burn, it is a lot faster than plasma, but it doesn't reach as high of a temperature as plasma (new somewhere over 70,000K, plasma over 90,000K, and old over 700,000K) when the reaction finishes.

CCC23 commented:
I hate this pr over anything else in the world

Pickle-Coding commented:
Why?

CCC23 commented:
I know its more the fact it screwed with alot of tritium combustion leading it to burn more inconsistently. IE trit for some reason burns cold along with plasma.

Pickle-Coding commented:
What do you mean burning cold?

CCC23 commented:
378 degree burn.

CCC23 commented:
378 burn while making trit

Pickle-Coding commented:
Because the burn changed, the optimal way to do it also changed. Re-read the atmos wiki for tritium synthesis for more information.
